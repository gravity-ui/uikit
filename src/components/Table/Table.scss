@use '../../../styles/mixins';
@use './variables';

#{variables.$block} {
    overflow: auto;
    position: relative;

    &__scroll-container {
        overflow: auto;
        position: relative;

        -ms-overflow-style: none;
        scrollbar-width: none;

        &::-webkit-scrollbar {
            display: none;
        }
    }

    &__horizontal-scroll-bar {
        overflow-x: auto;
        margin-top: -1px;

        &-inner {
            height: 1px;
            position: relative;

            // Safari 13 does not support transparent elements
            &::before {
                content: '';
                position: absolute;
                width: 1px;
                height: 1px;
                top: 0;
                left: 0;
                background-color: rgba(255, 255, 255, 0.01);
            }
        }

        &_sticky-horizontal-scroll {
            position: sticky;
            z-index: 3;
        }
    }

    &__table {
        border-spacing: 0;
        // fix border disappear in Firefox:
        border-collapse: separate;
    }

    &__cell {
        @include mixins.overflow-ellipsis();

        position: relative;
        box-sizing: content-box;
        padding: 11px 10px;
        border-bottom: 1px solid var(--g-color-line-generic);

        line-height: variables.$tableLineHeight;
        text-align: left;
        overflow-wrap: break-word;

        &:first-child {
            padding-left: 0;
        }

        &:last-child {
            padding-right: 0;
        }

        &:not(&_word-wrap) {
            @include mixins.overflow-ellipsis();
        }

        &_align_center {
            text-align: center;
        }

        &_align_right {
            text-align: right;
        }

        &_sticky_left,
        &_sticky_right {
            position: sticky;
            z-index: 2;
        }

        &_sticky_left,
        &_sticky_right {
            background: var(--g-color-base-background);
        }

        &_border_right {
            border-right: 1px solid var(--g-color-line-generic);
        }

        &_edge-padding {
            &:first-child {
                padding-left: 15px;
            }

            &:last-child {
                padding-right: 15px;
            }
        }
    }

    &__row {
        &_vertical-align_top {
            vertical-align: top;
        }

        &_vertical-align_middle {
            vertical-align: middle;
        }
    }

    &__row_empty &__cell {
        text-align: center;
    }

    &__body &__row:last-child &__cell {
        border-bottom-color: transparent;
    }

    &__th-content {
        @include mixins.text-accent;
        display: inline-block;
        &::first-letter {
            text-transform: uppercase;
        }
    }

    &__body &__row_interactive:hover {
        background-color: var(--g-color-base-simple-hover-solid);
        cursor: pointer;

        #{variables.$block}__cell_sticky_left,
        #{variables.$block}__cell_sticky_right {
            background: var(--g-color-base-simple-hover-solid);
        }
    }

    &__body &__row_disabled {
        opacity: 0.3;
    }

    &_with-primary &__body &__cell {
        color: var(--g-color-text-secondary);

        &_primary {
            color: var(--g-color-text-primary);
        }
    }

    // forcing overflow:visible so we are not breaking sticky scroll
    // (because sticky scroll relates to closest parent with overflow)
    &_with-sticky-scroll {
        overflow: visible;
    }
}
