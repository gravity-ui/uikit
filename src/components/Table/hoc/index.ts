export {withTableSelection} from './withTableSelection/withTableSelection';
export type {WithTableSelectionProps} from './withTableSelection/withTableSelection';
export {withTableActions} from './withTableActions/withTableActions';
export type {
    RenderRowActionsProps,
    WithTableActionsProps,
    TableActionConfig,
    TableAction,
    TableActionGroup,
} from './withTableActions/withTableActions';
export {withTableCopy} from './withTableCopy/withTableCopy';
export type {WithTableCopyProps} from './withTableCopy/withTableCopy';
export {withTableSorting, TableSortIndicator} from './withTableSorting/withTableSorting';
export type {
    WithTableSortingProps,
    TableSortState,
    TableColumnSortState,
} from './withTableSorting/withTableSorting';
export {withTableSettings} from './withTableSettings/withTableSettings';
export type {
    WithTableSettingsProps,
    TableSettingsData,
    TableColumnSetupItem,
} from './withTableSettings/withTableSettings';
export * from './withTableSettings/TableColumnSetup/TableColumnSetup';
