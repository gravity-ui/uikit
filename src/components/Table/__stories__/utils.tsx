import {withTableSelection} from '..';
import {Table} from '../Table';
import type {TableColumnConfig} from '../Table';
import {withTableActions, withTableCopy, withTableSettings, withTableSorting} from '../hoc';

export interface DataItem {
    name: string;
    city?: string;
    phone: string;
    count: number;
    date: string;
    disabled?: boolean;
}

export const data: DataItem[] = [
    {
        name: 'Nomlanga Compton',
        city: 'Erli',
        phone: '+7 (923) 737-89-72',
        count: 82,
        date: '2019-03-15',
    },
    {
        name: 'Paul Hatfield',
        city: 'Campitello di Fassa',
        phone: '+7 (900) 333-82-02',
        count: 51,
        date: '2019-11-23',
    },
    {
        name: 'Phelan Daniel',
        city: 'Meugliano',
        phone: '+7 (925) 549-50-23',
        count: 10,
        date: '2019-05-14',
    },
    {
        name: 'Hiram Mayer',
        city: '',
        phone: '+7 (950) 372-56-84',
        count: 54,
        date: '2019-03-29',
    },
    {
        name: 'Madeline Puckett',
        phone: '+7 (908) 582-05-91',
        count: 75,
        date: '2019-02-01',
        disabled: true,
    },
];

export const columns: TableColumnConfig<DataItem>[] = [
    {
        id: 'name',
        name: 'Name',
    },
    {
        id: 'city',
        name: 'City',
    },
    {
        id: 'phone',
        name: 'Phone',
    },
    {
        id: 'count',
        name: 'Count',
        align: 'right',
    },
    {
        id: 'date',
        name: 'Date created',
    },
];

export const TableWithAction = withTableActions<DataItem>(Table);
export const TableWithCopy = withTableCopy<DataItem>(Table);
export const TableWithSelection = withTableSelection<DataItem>(Table);
export const TableWithSettings = withTableSettings<DataItem>(Table);
export const TableWithSorting = withTableSorting<DataItem>(Table);
