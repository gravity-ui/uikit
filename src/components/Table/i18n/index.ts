import {addComponentKeysets} from '../../utils/addComponentKeysets';

import en from './en.json';
import ru from './ru.json';

const COMPONENT = 'Table';

export default addComponentKeysets({en, ru}, COMPONENT);
