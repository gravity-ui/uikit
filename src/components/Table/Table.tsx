'use client';

import React from 'react';

import _get from 'lodash/get';
import _has from 'lodash/has';
import _isNumber from 'lodash/isNumber';

import type {QAProps} from '../types';
import {block} from '../utils/cn';
import {warnOnce} from '../utils/warn';

import i18n from './i18n';

import './Table.scss';

const DASH = '\u2014';

export interface TableDataItem {
    [key: string]: any;
}

type ActiveScrollElementType = 'scrollBar' | 'scrollContainer';

function warnAboutPhysicalValues(propName: string) {
    warnOnce(
        `[Table] Physical values (left, right) of "${propName}" property are deprecated. Use logical values (start, end) instead.`,
    );
}

function normalizeSides(
    side: TableColumnConfig<any>['align'] | TableColumnConfig<any>['sticky'],
    propName: string,
) {
    if (side === 'left') {
        warnAboutPhysicalValues(propName);
        return 'start';
    }
    if (side === 'right') {
        warnAboutPhysicalValues(propName);
        return 'end';
    }
    return side;
}

interface TableState {
    // activeScrollElement is required so listener on table scroll won't fire when scrollbar will appear (and vice-versa)
    // without that page will wobble on scrolling
    activeScrollElement: ActiveScrollElementType;
    columnsStyles: React.CSSProperties[];
    columnHeaderRefs: React.RefObject<HTMLTableCellElement>[];
}

export interface TableColumnConfig<I> {
    /** Column ID */
    id: string;
    /** Column name (header). By default: column ID */
    name?: string | (() => React.ReactNode);
    /** CSS-class that will be added to all cells in the column. */
    className?: string;
    /** Stub in the event there is no data in a cell. By default: â€” (&mdash;) */
    placeholder?: string | ((item: I, index: number) => React.ReactNode);
    /** Cell contents. If you pass a row, the cell contents will be the value of the field named the same as this row. By default: The value of the field with the name equal to the column ID */
    template?: string | ((item: I, index: number) => React.ReactNode);
    /** Content alignment. */
    align?: 'start' | 'end' | 'center' | 'left' | 'right';
    /** Sticky column. */
    sticky?: 'start' | 'end' | 'left' | 'right';
    /** Distinguishes a column among other. */
    primary?: boolean;
    /** Column width in px or in %. Width can behave unexpectedly (it's more like min-width in block-elements). Sometimes you want to use `table-layout: fixed` */
    width?: number | string;
    /** Various data, HOC settings. */
    meta?: Record<string, any>;
}

export interface DescriptorType {
    /**
     * Row ID.
     * Used when selecting and sorting rows.
     */
    id?: string;

    /**
     * Row CSS classes.
     */
    classNames?: string[];

    /**
     * Condition for disabling columns.
     */
    disabled?: boolean;
}

// TODO: Replace @default in props description with defaultProps in order to work with Storybook.
export interface TableProps<I> extends QAProps {
    /** Data */
    data: I[];
    /** Column parameters */
    columns: TableColumnConfig<I>[];
    /** Vertical alignment of contents  */
    verticalAlign?: 'top' | 'middle';
    /** Break long text to lines instead of cutting with hellip */
    wordWrap?: boolean;
    /**
     * Horizontal sticky scroll.
     * Note: table cannot be with fixed height and with sticky scroll at the same time.
     * Sticky scroll wont work if table has overflow.
     *
     * @default false
     */
    stickyHorizontalScroll?: boolean;
    /**
     * Threshold when sticky scroll is enabled.
     *
     *  @default 0
     */
    stickyHorizontalScrollBreakpoint?: number;
    /**
     * @deprecated Use getRowDescriptor instead
     *
     * Row ID.
     * Used when selecting and sorting rows. If you pass a row,
     * its ID will be the value of the field in the row data named the same as the column ID.
     */
    getRowId?: string | ((item: I, index: number) => string);
    /**
     * @deprecated Use getRowDescriptor instead
     *
     * Row CSS classes.
     * */
    getRowClassNames?: (item: I, index: number) => string[];
    /**
     * @deprecated Use getRowDescriptor instead
     *
     * Condition for disabling columns.
     * */
    isRowDisabled?: (item: I, index: number) => boolean;

    /**
     *
     * @returns {DescriptorType} {@link DescriptorType}
     */
    getRowDescriptor?: (item: I, index: number) => DescriptorType | undefined;

    /** Row click handler. When passed row's hover is visible. */
    onRowClick?: (item: I, index: number, event: React.MouseEvent<HTMLTableRowElement>) => void;
    /** Row mouseenter handler. */
    onRowMouseEnter?: (
        item: I,
        index: number,
        event: React.MouseEvent<HTMLTableRowElement>,
    ) => void;
    /** Row mouseleave handler. */
    onRowMouseLeave?: (
        item: I,
        index: number,
        event: React.MouseEvent<HTMLTableRowElement>,
    ) => void;
    /** Row mousedown handler. */
    onRowMouseDown?: (item: I, index: number, event: React.MouseEvent<HTMLTableRowElement>) => void;
    /** Message returned if data is missing. By default: "No data". */
    emptyMessage?: string;
    /** Table CSS-class. */
    className?: string;
    /** Adds horizontal padding for edge cells. */
    edgePadding?: boolean;
}

interface TableDefaultProps {
    edgePadding: boolean;
}

const b = block('table');

export class Table<I extends TableDataItem = Record<string, string>> extends React.Component<
    TableProps<I>,
    TableState
> {
    static defaultProps: TableDefaultProps = {
        edgePadding: true,
    };

    // Static methods may be used by HOCs
    static getRowId<I extends TableDataItem>(props: TableProps<I>, item: I, rowIndex?: number) {
        const {data, getRowId, getRowDescriptor} = props;
        const index = rowIndex ?? data.indexOf(item);

        const descriptor = getRowDescriptor?.(item, index);

        if (descriptor?.id !== undefined) {
            return descriptor.id;
        }

        if (typeof getRowId === 'function') {
            return getRowId(item, index);
        }

        if (getRowId && getRowId in item) {
            return String(item[getRowId]);
        }

        return String(index);
    }

    static getHeadCellContent<I extends TableDataItem>(
        column: TableColumnConfig<I>,
    ): React.ReactNode {
        const {id, name} = column;

        let content;
        if (typeof name === 'function') {
            content = name();
        } else if (typeof name === 'string') {
            content = name;
        } else {
            content = id;
        }

        return content;
    }

    static getBodyCellContent<I extends TableDataItem>(
        column: TableColumnConfig<I>,
        item: I,
        rowIndex: number,
    ): React.ReactNode {
        const {id, template, placeholder} = column;

        let placeholderValue;
        if (typeof placeholder === 'function') {
            placeholderValue = placeholder(item, rowIndex);
        } else {
            placeholderValue = placeholder ?? DASH;
        }

        let value;
        if (typeof template === 'function') {
            value = template(item, rowIndex);
        } else if (typeof template === 'string') {
            value = _get(item, template);
        } else if (_has(item, id)) {
            value = _get(item, id);
        }
        if ([undefined, null, ''].includes(value as any) && placeholderValue) {
            return placeholderValue;
        }

        return value;
    }

    static getDerivedStateFromProps<I extends TableDataItem>(
        props: Readonly<TableProps<I>>,
        state: Readonly<TableState>,
    ): Partial<typeof state> | null {
        if (props.columns.length === state.columnHeaderRefs.length) {
            return null;
        }

        return {
            columnHeaderRefs: Array.from(props.columns, () => React.createRef()),
        };
    }

    state: TableState = {
        activeScrollElement: 'scrollContainer',
        columnsStyles: Array.from(this.props.columns, () => ({})),
        columnHeaderRefs: Array.from(this.props.columns, () => React.createRef()),
    };

    private tableRef = React.createRef<HTMLTableElement>();
    private scrollContainerRef = React.createRef<HTMLDivElement>();
    private horizontalScrollBarRef = React.createRef<HTMLDivElement>();
    private horizontalScrollBarInnerRef = React.createRef<HTMLDivElement>();

    private tableResizeObserver: ResizeObserver | undefined;
    private columnsResizeObserver: ResizeObserver | undefined;

    componentDidMount() {
        if (this.props.stickyHorizontalScroll) {
            this.tableResizeObserver = new ResizeObserver((entries: ResizeObserverEntry[]) => {
                const {contentRect} = entries[0];

                // Sync scrollbar width with table width
                this.horizontalScrollBarInnerRef.current?.style.setProperty(
                    'width',
                    `${contentRect.width}px`,
                );
            });

            if (this.tableRef.current) {
                this.tableResizeObserver.observe(this.tableRef.current);
            }

            if (this.scrollContainerRef.current) {
                this.scrollContainerRef.current.addEventListener(
                    'scroll',
                    this.handleScrollContainerScroll,
                );
                this.scrollContainerRef.current.addEventListener(
                    'mouseenter',
                    this.handleScrollContainerMouseenter,
                );
            }

            if (this.horizontalScrollBarRef.current) {
                this.horizontalScrollBarRef.current.addEventListener(
                    'scroll',
                    this.handleHorizontalScrollBarScroll,
                );
                this.horizontalScrollBarRef.current.addEventListener(
                    'mouseenter',
                    this.handleHorizontalScrollBarMouseenter,
                );
            }
        }

        this.columnsResizeObserver = new ResizeObserver((entries) => {
            // fix ResizeObserver loop error
            window.requestAnimationFrame(() => {
                if (!Array.isArray(entries) || !entries.length) {
                    return;
                }
                this.updateColumnStyles();
            });
        });

        if (this.tableRef.current) {
            this.columnsResizeObserver.observe(this.tableRef.current);
        }

        this.updateColumnStyles();
    }

    componentDidUpdate(prevProps: TableProps<I>) {
        if (this.props.columns !== prevProps.columns) {
            this.updateColumnStyles();
        }
    }

    componentWillUnmount() {
        if (this.props.stickyHorizontalScroll) {
            if (this.tableResizeObserver) {
                this.tableResizeObserver.disconnect();
            }

            if (this.scrollContainerRef.current) {
                this.scrollContainerRef.current.removeEventListener(
                    'scroll',
                    this.handleScrollContainerScroll,
                );
                this.scrollContainerRef.current.removeEventListener(
                    'mouseenter',
                    this.handleScrollContainerMouseenter,
                );
            }

            if (this.horizontalScrollBarRef.current) {
                this.horizontalScrollBarRef.current.removeEventListener(
                    'scroll',
                    this.handleHorizontalScrollBarScroll,
                );
                this.horizontalScrollBarRef.current.removeEventListener(
                    'mouseenter',
                    this.handleHorizontalScrollBarMouseenter,
                );
            }
        }

        if (this.columnsResizeObserver) {
            this.columnsResizeObserver.disconnect();
        }
    }

    render() {
        const {columns, stickyHorizontalScroll, className, qa} = this.props;
        const withPrimary = columns.some(({primary}) => primary);

        return (
            <div
                className={b(
                    {
                        'with-primary': withPrimary,
                        'with-sticky-scroll': stickyHorizontalScroll,
                    },
                    className,
                )}
                data-qa={qa}
            >
                {stickyHorizontalScroll ? (
                    <React.Fragment>
                        <div ref={this.scrollContainerRef} className={b('scroll-container')}>
                            {this.renderTable()}
                        </div>
                        {this.renderHorizontalScrollBar()}
                    </React.Fragment>
                ) : (
                    this.renderTable()
                )}
            </div>
        );
    }

    private renderHead() {
        const {columns, edgePadding, wordWrap} = this.props;
        const {columnsStyles} = this.state;

        return (
            <thead className={b('head')}>
                <tr className={b('row')}>
                    {columns.map((column, index) => {
                        const {id, align: rawAlign, primary, sticky: rawSticky, className} = column;
                        const align = normalizeSides(rawAlign, 'column.align');
                        const sticky = normalizeSides(rawSticky, 'column.sticky');
                        const content = Table.getHeadCellContent(column);

                        return (
                            <th
                                key={id}
                                ref={this.state.columnHeaderRefs[index]}
                                style={columnsStyles[index]}
                                className={b(
                                    'cell',
                                    {
                                        align,
                                        primary,
                                        sticky,
                                        ['edge-padding']: edgePadding,
                                        ['word-wrap']: wordWrap,
                                    },
                                    className,
                                )}
                            >
                                {content}
                            </th>
                        );
                    })}
                </tr>
            </thead>
        );
    }

    private renderBody() {
        const {data} = this.props;

        return (
            <tbody className={b('body')}>
                {data.length > 0 ? data.map(this.renderRow) : this.renderEmptyRow()}
            </tbody>
        );
    }

    private renderTable() {
        return (
            <table ref={this.tableRef} className={b('table')}>
                {this.renderHead()}
                {this.renderBody()}
            </table>
        );
    }

    private renderRow = (item: I, rowIndex: number) => {
        const {
            columns,
            isRowDisabled,
            onRowClick,
            onRowMouseEnter,
            onRowMouseLeave,
            onRowMouseDown,
            getRowClassNames,
            verticalAlign,
            edgePadding,
            wordWrap,
            getRowDescriptor,
        } = this.props;
        const {columnsStyles} = this.state;

        const descriptor = getRowDescriptor?.(item, rowIndex);

        const disabled = descriptor?.disabled || isRowDisabled?.(item, rowIndex) || false;

        const additionalClassNames =
            descriptor?.classNames || getRowClassNames?.(item, rowIndex) || [];

        const interactive = Boolean(!disabled && onRowClick);

        return (
            <tr
                key={Table.getRowId(this.props, item, rowIndex)}
                onClick={
                    !disabled && onRowClick ? onRowClick.bind(null, item, rowIndex) : undefined
                }
                onMouseEnter={
                    !disabled && onRowMouseEnter
                        ? onRowMouseEnter.bind(null, item, rowIndex)
                        : undefined
                }
                onMouseLeave={
                    !disabled && onRowMouseLeave
                        ? onRowMouseLeave.bind(null, item, rowIndex)
                        : undefined
                }
                onMouseDown={
                    !disabled && onRowMouseDown
                        ? onRowMouseDown.bind(null, item, rowIndex)
                        : undefined
                }
                className={b(
                    'row',
                    {disabled, interactive, 'vertical-align': verticalAlign},
                    additionalClassNames.join(' '),
                )}
            >
                {columns.map((column, colIndex) => {
                    const {id, align: rawAlign, primary, className, sticky: rawSticky} = column;
                    const content = Table.getBodyCellContent(column, item, rowIndex);
                    const align = normalizeSides(rawAlign, 'column.align');
                    const sticky = normalizeSides(rawSticky, 'column.sticky');
                    return (
                        <td
                            key={id}
                            style={columnsStyles[colIndex]}
                            className={b(
                                'cell',
                                {
                                    align,
                                    primary,
                                    sticky,
                                    ['edge-padding']: edgePadding,
                                    ['word-wrap']: wordWrap,
                                },
                                className,
                            )}
                        >
                            {content}
                        </td>
                    );
                })}
            </tr>
        );
    };

    private renderEmptyRow() {
        const {columns, emptyMessage} = this.props;

        return (
            <tr className={b('row', {empty: true})}>
                <td className={b('cell')} colSpan={columns.length}>
                    {emptyMessage ? emptyMessage : i18n('label_empty')}
                </td>
            </tr>
        );
    }

    private renderHorizontalScrollBar() {
        const {stickyHorizontalScroll, stickyHorizontalScrollBreakpoint = 0} = this.props;
        return (
            <div
                ref={this.horizontalScrollBarRef}
                className={b('horizontal-scroll-bar', {
                    'sticky-horizontal-scroll': stickyHorizontalScroll,
                })}
                style={{bottom: `${stickyHorizontalScrollBreakpoint}px`}}
                data-qa="sticky-horizontal-scroll-breakpoint-qa"
            >
                <div
                    ref={this.horizontalScrollBarInnerRef}
                    className={b('horizontal-scroll-bar-inner')}
                />
            </div>
        );
    }

    private updateColumnStyles() {
        this.setState((prevState) => {
            const columnsWidth = prevState.columnHeaderRefs.map((ref) =>
                ref.current === null ? undefined : ref.current.getBoundingClientRect().width,
            );

            const columnsStyles = this.props.columns.map((_, index) =>
                this.getColumnStyles(index, columnsWidth),
            );

            return {columnsStyles};
        });
    }

    private getColumnStyles(index: number, columnsWidth: (number | undefined)[]) {
        const {columns} = this.props;

        const column = columns[index];
        const style: React.CSSProperties = {};
        if (typeof column.width === 'string') {
            return {maxWidth: 0, width: column.width};
        }
        if (typeof column.width !== 'undefined') {
            style.width = column.width;
        }
        if (!column.sticky) {
            return style;
        }

        const filteredColumns =
            column.sticky === 'left' || column.sticky === 'start'
                ? columnsWidth.slice(0, index)
                : columnsWidth.slice(index + 1);
        const styleName: keyof React.CSSProperties =
            column.sticky === 'left' || column.sticky === 'start'
                ? 'insetInlineStart'
                : 'insetInlineEnd';
        style[styleName] = filteredColumns.reduce<number>((start, width) => {
            return _isNumber(width) ? start + width : start;
        }, 0);

        return style;
    }

    private handleScrollContainerMouseenter = () => {
        this.setState({activeScrollElement: 'scrollContainer'});
    };

    private handleScrollContainerScroll = () => {
        if (
            this.state.activeScrollElement === 'scrollContainer' &&
            this.horizontalScrollBarRef.current &&
            this.scrollContainerRef.current
        ) {
            this.horizontalScrollBarRef.current.scrollLeft =
                this.scrollContainerRef.current.scrollLeft;
        }
    };

    private handleHorizontalScrollBarMouseenter = () => {
        this.setState({activeScrollElement: 'scrollBar'});
    };

    private handleHorizontalScrollBarScroll = () => {
        if (
            this.state.activeScrollElement === 'scrollBar' &&
            this.horizontalScrollBarRef.current &&
            this.scrollContainerRef.current
        ) {
            this.scrollContainerRef.current.scrollLeft =
                this.horizontalScrollBarRef.current.scrollLeft;
        }
    };
}
