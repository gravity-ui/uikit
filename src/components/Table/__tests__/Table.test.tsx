import React from 'react';

import {render, screen, within} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import {Table} from '../Table';

import {columns, data} from './utils';

const qaId = 'table-component';

describe('Table', () => {
    test('render table by default', () => {
        render(<Table data={data} columns={columns} />);
        const table = screen.getByRole('table');
        const rows = within(table).getAllByRole('row');

        expect(table).toBeVisible();
        expect(rows.length).toBe(data.length + 1);
    });

    test.each(new Array<'top' | 'middle'>('top', 'middle'))(
        'render with given "%s" vertical align',
        (align) => {
            render(<Table data={data} columns={columns} verticalAlign={align} />);
            const table = screen.getByRole('table');
            const tbody = within(table).getAllByRole('rowgroup');
            const rows = within(tbody[1]).getAllByRole('row');

            rows.forEach((row) => {
                expect(row).toHaveClass(`yc-table__row_vertical-align_${align}`);
            });
        },
    );

    test('render table with no data (default)', () => {
        const emptyText = 'No data';
        render(<Table data={[]} columns={columns} />);
        const emptyCell = screen.getByText(emptyText);

        expect(emptyCell).not.toBeNull();
    });

    test('render table with no data (custom)', () => {
        const emptyText = 'Custom no data message';
        render(<Table data={[]} columns={columns} emptyMessage={emptyText} />);
        const emptyCell = screen.getByText(emptyText);

        expect(emptyCell).not.toBeNull();
    });

    test('render table with edge padding', () => {
        render(<Table data={data} columns={columns} edgePadding={true} />);
        const table = screen.getByRole('table');
        const thead = within(table).getAllByRole('rowgroup');
        const cells = within(thead[0]).getAllByRole('columnheader');

        cells.forEach((cell) => {
            expect(cell).toHaveClass(`yc-table__cell_edge-padding`);
        });
    });

    test('render table without edge padding', () => {
        render(<Table data={data} columns={columns} edgePadding={false} />);
        const table = screen.getByRole('table');
        const thead = within(table).getAllByRole('rowgroup');
        const cells = within(thead[0]).getAllByRole('columnheader');

        cells.forEach((cell) => {
            expect(cell).not.toHaveClass(`yc-table__cell_edge-padding`);
        });
    });

    test('render table with disabled row', () => {
        render(
            <Table
                data={data}
                columns={columns}
                isRowDisabled={(item) => Boolean(item.disabled)}
            />,
        );
        const table = screen.getByRole('table');
        const tbody = within(table).getAllByRole('rowgroup');
        const rows = within(tbody[1]).getAllByRole('row');

        const row = rows.filter((r) => r.className.includes('yc-table__row_disabled'));
        expect(row.length).toBe(1);
    });

    test('render table with custrom row className', () => {
        render(
            <Table
                data={data}
                columns={columns}
                getRowClassNames={(_, index) => [`my-class-${index}`]}
            />,
        );
        const table = screen.getByRole('table');
        const tbody = within(table).getAllByRole('rowgroup');
        const rows = within(tbody[1]).getAllByRole('row');

        rows.forEach((row, index) =>
            expect(row.className.includes(`my-class-${index}`)).toBe(true),
        );
    });

    test('add className', () => {
        const className = 'my-class';

        render(<Table data={data} columns={columns} className={className} qa={qaId} />);
        const table = screen.getByTestId(qaId);

        expect(table).toHaveClass(className);
    });

    test('render table with horizontal sticky scroll', () => {
        render(<Table data={data} columns={columns} stickyHorizontalScroll={true} qa={qaId} />);
        const table = screen.getByTestId(qaId);

        expect(table).toHaveClass('yc-table_with-sticky-scroll');
    });

    test('call onRowClick', async () => {
        const user = userEvent.setup();
        const handleOnClick = jest.fn();
        render(<Table data={data} columns={columns} onRowClick={handleOnClick} />);

        const table = screen.getByRole('table');
        const tbody = within(table).getAllByRole('rowgroup');
        const rows = within(tbody[1]).getAllByRole('row');
        const promises = rows.map(async (row) => {
            await user.click(row);
        });
        await Promise.all(promises);

        expect(handleOnClick).toHaveBeenCalledTimes(data.length);
    });

    test('render table with horizontal sticky scroll and custom breakpoint', () => {
        const stickyHorizontalScrollBreakpointQa = 'sticky-horizontal-scroll-breakpoint-qa';
        const stickyHorizontalScrollBreakpoint = 777;
        const style = {bottom: `${stickyHorizontalScrollBreakpoint}px`};
        render(
            <Table
                data={data}
                columns={columns}
                stickyHorizontalScroll={true}
                stickyHorizontalScrollBreakpoint={stickyHorizontalScrollBreakpoint}
            />,
        );
        const table = screen.getByTestId(stickyHorizontalScrollBreakpointQa);

        expect(table).toHaveStyle(style);
    });
});
