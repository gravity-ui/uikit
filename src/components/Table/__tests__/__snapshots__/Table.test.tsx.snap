// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should correctly add custom row classnames 1`] = `
<div>
  <div
    class="yc-table"
  >
    <table
      class="yc-table__table"
    >
      <thead
        class="yc-table__head"
      >
        <tr
          class="yc-table__row"
        >
          <th
            class="yc-table__cell yc-table__cell_edge-padding"
          >
            name
          </th>
        </tr>
      </thead>
      <tbody
        class="yc-table__body"
      >
        <tr
          class="yc-table__row custom-row custom-row_mod"
        >
          <td
            class="yc-table__cell yc-table__cell_edge-padding"
          >
            —
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should correctly add disabled classname 1`] = `
<div>
  <div
    class="yc-table"
  >
    <table
      class="yc-table__table"
    >
      <thead
        class="yc-table__head"
      >
        <tr
          class="yc-table__row"
        >
          <th
            class="yc-table__cell yc-table__cell_edge-padding"
          >
            name
          </th>
        </tr>
      </thead>
      <tbody
        class="yc-table__body"
      >
        <tr
          class="yc-table__row"
        >
          <td
            class="yc-table__cell yc-table__cell_edge-padding"
          >
            —
          </td>
        </tr>
        <tr
          class="yc-table__row yc-table__row_disabled"
        >
          <td
            class="yc-table__cell yc-table__cell_edge-padding"
          >
            —
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should correctly apply \`align\` prop 1`] = `
<div>
  <div
    class="yc-table"
  >
    <table
      class="yc-table__table"
    >
      <thead
        class="yc-table__head"
      >
        <tr
          class="yc-table__row"
        >
          <th
            class="yc-table__cell yc-table__cell_align_left yc-table__cell_edge-padding"
          >
            name1
          </th>
          <th
            class="yc-table__cell yc-table__cell_align_right yc-table__cell_edge-padding"
          >
            name2
          </th>
          <th
            class="yc-table__cell yc-table__cell_align_center yc-table__cell_edge-padding"
          >
            name3
          </th>
        </tr>
      </thead>
      <tbody
        class="yc-table__body"
      >
        <tr
          class="yc-table__row"
        >
          <td
            class="yc-table__cell yc-table__cell_align_left yc-table__cell_edge-padding"
          >
            —
          </td>
          <td
            class="yc-table__cell yc-table__cell_align_right yc-table__cell_edge-padding"
          >
            —
          </td>
          <td
            class="yc-table__cell yc-table__cell_align_center yc-table__cell_edge-padding"
          >
            —
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should correctly apply \`primary\` prop 1`] = `
<div>
  <div
    class="yc-table yc-table_with-primary"
  >
    <table
      class="yc-table__table"
    >
      <thead
        class="yc-table__head"
      >
        <tr
          class="yc-table__row"
        >
          <th
            class="yc-table__cell yc-table__cell_primary yc-table__cell_edge-padding"
          >
            name1
          </th>
          <th
            class="yc-table__cell yc-table__cell_edge-padding"
          >
            name2
          </th>
          <th
            class="yc-table__cell yc-table__cell_edge-padding"
          >
            name3
          </th>
        </tr>
      </thead>
      <tbody
        class="yc-table__body"
      >
        <tr
          class="yc-table__row"
        >
          <td
            class="yc-table__cell yc-table__cell_primary yc-table__cell_edge-padding"
          >
            —
          </td>
          <td
            class="yc-table__cell yc-table__cell_edge-padding"
          >
            —
          </td>
          <td
            class="yc-table__cell yc-table__cell_edge-padding"
          >
            —
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should correctly apply \`width\` prop 1`] = `
<div>
  <div
    class="yc-table"
  >
    <table
      class="yc-table__table"
    >
      <thead
        class="yc-table__head"
      >
        <tr
          class="yc-table__row"
        >
          <th
            class="yc-table__cell yc-table__cell_edge-padding"
            style="width: 100px;"
          >
            name1
          </th>
          <th
            class="yc-table__cell yc-table__cell_edge-padding"
            style="width: 200px;"
          >
            name2
          </th>
          <th
            class="yc-table__cell yc-table__cell_edge-padding"
            style="width: 300px;"
          >
            name3
          </th>
        </tr>
      </thead>
      <tbody
        class="yc-table__body"
      >
        <tr
          class="yc-table__row"
        >
          <td
            class="yc-table__cell yc-table__cell_edge-padding"
            style="width: 100px;"
          >
            —
          </td>
          <td
            class="yc-table__cell yc-table__cell_edge-padding"
            style="width: 200px;"
          >
            —
          </td>
          <td
            class="yc-table__cell yc-table__cell_edge-padding"
            style="width: 300px;"
          >
            —
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should render empty state 1`] = `
<div>
  <div
    class="yc-table"
  >
    <table
      class="yc-table__table"
    >
      <thead
        class="yc-table__head"
      >
        <tr
          class="yc-table__row"
        >
          <th
            class="yc-table__cell yc-table__cell_edge-padding"
          >
            name
          </th>
        </tr>
      </thead>
      <tbody
        class="yc-table__body"
      >
        <tr
          class="yc-table__row yc-table__row_empty"
        >
          <td
            class="yc-table__cell"
            colspan="1"
          >
            No data
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
